File: Heap Sort/Heap Sort.cpp
/// This code is contributed by @notreallystatic

/// Time Complexity: O(nlogn)
/// Space Complexity: O(1) (Arrange the array into a heap, in place.)

// Using vectors to represent an array.
#include <iostream>
#include <vector>

using namespace std;

int left(const int &i)
{
  return (2*i+1);
}

int right(const int &i)
{
  return (2*i+2);
}

void heapify(vector<int> &arr, const int &i, const int &n)
{
  int l=left(i);
  int r=right(i);
  int largest=i;
  if(l<n && arr[l]>arr[i])
    largest=l;
  if(r<n && arr[r]>arr[largest])
    largest=r;
  if(largest!=i)
  {
    swap(arr[largest], arr[i]);
    heapify(arr, largest, n);
  }
}

void heapSort(vector<int> &arr)
{
  int n=arr.size();

  for(int i=n/2-1; i>=0; i--) {
    heapify(arr, i, n);
  }

  for(int i=n-1; i>=0; i--) {
    swap(arr[i], arr[0]);
    heapify(arr, 0, i);
  }
}

int main()
{

  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NUll);

  vector<int> arr = {0, 3, 10, -2, 20, 19, 2};

  heapSort(arr);

  for(int x : arr){
    cout << x << " ";
  }

  return 0;
}

File: Heap Sort/Heap Sort.py
# This code is contributed by @iuliagroza

# Time Complexity: O(nlogn)
# Space Complexity: O(1) // Arrange the array into a heap, in place.

def heapify(arr, n, i):
    largest = i  # Initialize largest as root
    l = 2*i+1
    r = 2*i+2

    # See if left child of root exists and is greater than root
    if l < n and arr[largest] < arr[l]:
        largest = l

    # See if right child of root exists and is greater than root
    if r < n and arr[largest] < arr[r]:
        largest = r

    # Change root, if needed
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # swap
        # Heapify the root.
        heapify(arr, n, largest)

def heapSort(arr):
    n = len(arr)
    # Build a maxheap.
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # One by one extract elements
    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # swap
        heapify(arr, i, 0)

arr = [12, 11, 13, 5, 6, 7]
heapSort(arr)
n = len(arr)
for i in range(n):
    print("%d" % arr[i])